synapse/lib/json.py

```diff
@@ -0,0 +1,191 @@
+            ret = ret.decode()
+        return ret
+    except orjson.JSONEncodeError as exc:
+        raise s_exc.MustBeJsonSafe(mesg=exc.args[0])
+
+def dump(obj, fp, sort_keys=False, indent=False, default=None, append_newline=False):
+    '''
+    Serialize a python object to a string.
+
+    Similar to the standard library json.dump().
+
+    Note: This function attempts to automatically detect if the file was opened
+    in text or binary mode by inspecting `fp.mode`.
+
+    Arguments:
+        obj (object): The python object to serialize.
+        fp (file): The python file pointer to write the serialized data to.
+        sort_keys (Optional[bool]): Sort dictionary keys. Default: False.
+        indent (Optional[bool]): Include 2 spaces of indentation. Default: False.
+        default (Optional[callable]): Callback for serializing unknown object types. Default: None.
+        append_newlines (Optional[bool]): Append a newline to the end of the serialized data. Default: False.
+
+    Returns: None
+
+    Raises:
+        synapse.exc.MustBeJsonSafe: This exception is raised when a python object cannot be serialized.
+    '''
+    asbytes = False
+    if hasattr(fp, 'mode'):
+        asbytes = 'b' in fp.mode
+    elif isinstance(fp, io.BytesIO):
+        asbytes = True
+
+    data = dumps(obj, sort_keys=sort_keys, indent=indent, default=default, asbytes=asbytes, append_newline=append_newline)
+    fp.write(data)
+
+def jsload(*paths):
+    '''
+    Deserialize the JSON data at *paths.
+
+    Arguments:
+        *paths: The file path parts to load the data from.
+
+    Returns:
+        (object): The deserialized JSON data.
+
+    Raises:
+        synapse.exc.BadJsonText: This exception is raised when there is an error
+            deserializing the provided data.
+    '''
+    import synapse.common as s_common # Avoid circular import
+    with s_common.genfile(*paths) as fd:
+        if os.fstat(fd.fileno()).st_size == 0:
+            return None
+
+        return load(fd)
+
+def jslines(*paths):
+    '''
+    Deserialize the JSON lines data at *paths.
+
+    Arguments:
+        *paths: The file path parts to load the data from.
+
+    Yields:
+        (object): The deserialized JSON data from each line.
+
+    Raises:
+        synapse.exc.BadJsonText: This exception is raised when there is an error
+            deserializing the provided data.
+    '''
+    import synapse.common as s_common # Avoid circular import
+    with s_common.genfile(*paths) as fd:
+        for line in fd:
+            yield loads(line)
+
+def jssave(js, *paths):
+    '''
+    Serialize the python object to a file.
+
+    Arguments:
+        js: The python object to serialize.
+        *paths: The file path parts to save the data to.
+
+    Returns: None
+
+    Raises:
+        synapse.exc.MustBeJsonSafe: This exception is raised when a python
+        object cannot be serialized.
+    '''
+    import synapse.common as s_common # Avoid circular import
```

Comment:
i've also got one of these with some of the new logging stuff. Could we potentially further isolate the genpath type functions to avoid this?

Reply:
We could but it would probably break a ton of stuff. Happy to discuss.